src/execution/value_stack.rs

```diff
@@ -12,81 +13,174 @@ use crate::unreachable_validated;
 /// 3. Activations
 ///
 /// See <https://webassembly.github.io/spec/core/exec/runtime.html#stack>
-#[derive(Clone, Debug, Default)]
+#[derive(Default)]
 pub(crate) struct Stack {
-    inner: Vec<StackEntry>,
+    /// WASM values on the stack, i.e. the actual data that instructions operate on
+    values: Vec<Value>,
+
+    /// Stack frames
+    ///
+    /// Each time a function is called, a new frame is pushed, whenever a function returns, a frame is popped
+    frames: Vec<CallFrame>,
 }
 
 impl Stack {
     pub fn new() -> Self {
         Self::default()
     }
 
+    /// Pop a value of the given [ValType] from the value stack
     pub fn pop_value(&mut self, ty: ValType) -> Value {
-        let popped = self.inner.pop().unwrap_validated();
-        match popped {
-            StackEntry::Value(popped) => {
-                if popped.to_ty() == ty {
-                    popped
-                } else {
-                    unreachable_validated!()
-                }
-            }
-            StackEntry::Label(_) => todo!("pop label from stack"),
-            StackEntry::Activation(_) => todo!("pop activation frame from stack"),
+        // If there is at least one stack frame, we shall not pop values past the current
+        // stackframe. However, there is one legitimate reason to pop when there is **no** current
+        // stackframe: after the outermost function returns, to extract the final return values of
+        // this interpreter invocation.
+        debug_assert!(
```

Comment:
A question here, maybe we discussed about this, but would we want an error state/panic/etc. in the release mode if we manage to get here?

Reply:
Yes this is only there to find bugs when running `cargo test`. Basically, assuming that there are no logic bugs in the implementation, this will never panic.

Reply:
It did show me an intersting logic bug though: I inserted a `debug_assert` to verify that values are never popped from the value stack past the current stack frames base pointer. Turns out, there is a moment where there is 0 `CallFrame`s, but is still legal to pop values from the stack: after an invocation finished, to retrieve the return values.